### PR TITLE
feat(build): add Dagger based build module

### PR DESCRIPTION
**Pull request by GeorgeMac**
This adds a new Go `hack/build` directory and module used to build and test Flipt using Dagger.

There are more details in the README included in this PR.
However, as an overview, here are the added mage targets:
```sh
➜  cd build
➜  mage
Targets:
  build:base          builds Flipts base image via Dagger and buildkit.
  build:flipt         builds a development version of Flipt as a Docker image and loads it into a local Docker instance.
  test:all            runs Flipt's unit test suite against all the databases Flipt supports.
  test:database       runs the unit test suite against the desired database (one of ["sqlite" "postgres" "mysql" "cockroach"]).
  test:integration    runs the entire integration test suite.
  test:unit           runs the base suite of tests for all of Flipt.
```

For now, this is a fairly innocuous change. It doesn't change any of how CI works.
However, it lays some foundational build steps. Something we can iterate and experiment with further.

### CODE REVIEW

File: build/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+## Test
+
+The test section of the Mage targets handles running Flipts various unit tests with different configurations.
+
+`mage test:unit` runs the entire test suite with `SQLite` as the backing database.
+
+`mage test:database <db>` runs the entire test suite with the desired database (`sqlite`, `postgres`, `mysql` and `cockroach` available).
+
+`mage test:all` runs the entire suite for each database concurrently in parallel.
```
Comment (markphelps) [inline build/README.md:30]: amazing

File: build/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+`mage test:integration` run an [integration test suite](./build/integration) against an instance of Flipt.
```
Comment (markphelps) [inline build/README.md:32]:
```suggestion
`mage test:integration` run the [integration test suite](./build/integration) against an instance of Flipt.
```

File: build/internal/test/test.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	// create Redis service container
+	redisSrv := client.Container().
+		From("redis").
+		WithExposedPort(6379).
+		WithExec(nil)
+
+	_, err := flipt.
+		WithServiceBinding("redis", redisSrv).
+		WithEnvVariable("REDIS_HOST", "redis:6379").
+		WithExec([]string{"go", "test", "-tags", "assets", "-p", "1", "./..."}).
```
Comment (markphelps) [inline build/internal/test/test.go:19]: idk if we need `assets` tag here

File: build/internal/ui.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		}), nil
+	}
+
+	path, ref, _ := strings.Cut(path[len("git://"):], "#")
+	if ref == "" {
+		ref = "main"
+	}
+
+	socket := client.Host().UnixSocket(os.Getenv("SSH_AUTH_SOCK"))
+	return client.Git(path).
```
Comment (markphelps) [inline build/internal/ui.go:34]: does this do the clone?

File: build/internal/test/test.go (hunk 3 above)
Comment (GeorgeMac) [reply build/internal/test/test.go:19]: hmm yes I think you're right. As we're not exercising those paths within Flipts unit tests.

File: hack/build/internal/flipt.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+	// build the Flipt target binary
+	return golang.
+		WithMountedFile("./ui/embed.go", embed.File("./ui/embed.go")).
+		WithMountedDirectory("./ui/dist", req.ui).
+		WithExec([]string{"mkdir", "-p", req.binary()}).
+		WithExec([]string{"sh", "-c", goBuildCmd}), nil
+}
+
+// Package copies the Flipt binaries built into the provoded flipt container
```
Comment (yquansah) [inline hack/build/internal/flipt.go:187]:
```suggestion
// Package copies the Flipt binaries built into the provided flipt container
```

**Review by markphelps — APPROVED** [review]
lets go!

**Pull request merged**